### ISSUE TITLE
Use darknet to compare objects

### ISSUE THREAD

**Issue by Dinl** [AlexeyAB/darknet#2464]
Hi, I'm using Darknet to detect multiple objects in several frames, in order to maintain the track of each object I'm using several methods (mainly a high IoU and others); then, I have a general question, it is posible to train a darknet model to compare two detections and decide if they are same object?

**Comment by AlexeyAB** [comment]
@Dinl Hi,

Do you want to train a model that can detect objects on video with increased accuracy, based on the fact that each next frame is similar to the previous one?

**Comment by Dinl** [comment]
No, I already use the yolo3 model to detect with high accuracy, but the problem is with video and continuous detection: in some frames the objects are not detected. The idea is to train another model that compares the last frames detections and new detection in order to keep the track of the object.
The proposed input is 2 images of the objects to compare, and output is a single value for similarity, do you consider that this has sense with darknet?

**Comment by AlexeyAB** [comment]
Theoretically it can be implemented in the darknet, but there are 3 cases:

1. if there are occlusions (f.e. person cross another person), so the second person is invisible - in this case comparison of two images can't solve this problem

2. if there are occlusions (f.e. car cross another person), this case can be solved by using additional layers conv-rnn or conv-lstm, and several fixes in the training code

3. if there are not occlusions, in this case we can solve this problem just by using Optical Flow from OpenCV, that is already used in the `yolo_console_dll.cpp`

----

To use case-3:

Just un-comment this line: https://github.com/AlexeyAB/darknet/blob/f37ea75b28f01a87f08cb3cdea2f3d58f72224f8/src/yolo_console_dll.cpp#L14
build Darknet with `LIBSO=1` in the Makefile
and run 
`LD_LIBRARY_PATH=./:$LD_LIBRARY_PATH ./uselib data/coco.names cfg/yolov3.cfg yolov3.weights test.mp4`

If your OpenCV doesn't use CUDA, then change GPU to GPU_DISABLED here: https://github.com/AlexeyAB/darknet/blob/f37ea75b28f01a87f08cb3cdea2f3d58f72224f8/include/yolo_v2_class.hpp#L172

**Comment by Dinl** [comment]
Hi @AlexeyAB , still working on this, I have a initial prototype in Keras, there is a way to convert the weights from .h5 to .weights to load in darknet? 

I have been looking the [YAD2K](https://github.com/allanzelener/YAD2K) and [](https://github.com/Microsoft/MMdnn) but there are only in one way (from darknet to keras).

**Comment by AlexeyAB** [comment]
@Dinl Hi,

>  there is a way to convert the weights from .h5 to .weights to load in darknet?

I don't know about this.
